MdePkg/Include/Guid/DebugImageInfoTable.h

```diff
@@ -47,13 +48,55 @@ typedef struct {
   /// Indicates the image handle of the associated image.
   ///
   EFI_HANDLE                   ImageHandle;
-  CHAR8                      *PdbPath;
-  UINTN                        DebugBase;
 } EFI_DEBUG_IMAGE_INFO_NORMAL;
 
+typedef struct {
+  ///
+  /// When this debug image info structure is present, ImageInfoType is
+  /// set to EFI_DEBUG_IMAGE_INFO_TYPE_NORMAL2, indicating that PdbPath
+  /// and DebugBase fields are available for symbolication. The format
+  /// of the loaded image is not specified and may or may not be PE.
+  ///
+  UINT32                       ImageInfoType;
+  ///
+  /// The size of the current instance of this struct, including any
+  /// custom extensions.
+  ///
+  UINTN                        Size;

```

Comment:
FWIW I remain unconvinced by the size field (assuming I've implemented it even vaguely correctly). I'm coming at this from the PoV of 'what's going to require the least changes to the debugging code, moving forwards?' (Bearing in mind that GdbSyms.dll needs to be manually recompiled, in multiple versions done on at least two different OSes, when significant changes to the structures happen.)

Hence I'd prefer that EFI_DEBUG_IMAGE_INFO_TYPE_NORMAL2 takes the value 2 now and keeps it ... :-) And I'd prefer that the offset of PdbPath and DebugBase stays the same. Understood that recompilation _may_ be necessary in future, but I don't want to bake in the requirement for it. ... It seems to me that the mature version of this struct may well not have Size. It also _seems_ to me that Size does not help much. What if two different people make two different extensions to the struct, each of which is 8 bytes long (say two different pointers). Then size cannot let a consumer distinguish these. In fact, it seems that the producer and consumer are so tightly tied to having been compiled from the same code, in order for custom extensions to work, that there is no need for the Size field, for this.

And I believe that there is no real case for saving the size for copying and iterating. This struct lives in a context where, by definition, you do not know the size of the struct until you have read ImageInfoType - and once you have read it, you do know the size of the struct - it is the size of the struct (including any custom extensions which are currently there). Which will work, assuming we already have a pretty strict requirement that custom extensions only work if the producer and consumer are compiled from the same code - which, as above, I believe we do!

Reply:
You know the size from the type value *till you add more fields yet*, introduce type 3, and break the tools again. Whatever, drop it then.